File: crates/sc-preview/src/worker/jailed/seccomp.rs
//! The seccomp-BPF syscall allow-list for the preview worker
//! (`DESIGN-PREVIEW.md` §4.2, step ③).
//!
//! # Why the filter is hand-built instead of using a crate
//!
//! Three reasons, in order of weight:
//!
//! 1. **There is already a hand-built seccomp filter in this workspace.**
//!    `sc-server::hardening` installs a small deny-list for the server
//!    process itself (`ARCHITECTURE.md` §2.4) with the same `libc::sock_filter`
//!    primitives used here. Pulling in `seccompiler` for the second filter
//!    would leave the tree with two different mechanisms for the same job.
//! 2. **The policy is flat.** Twenty-two syscall numbers, no argument
//!    inspection, no ranges, no conditional actions. That is precisely the
//!    shape for which classic BPF is two dozen instructions of straight-line
//!    code — and it is the one place in the tree where "the entire policy the
//!    kernel will enforce is readable in one screen, with nothing in between"
//!    beats the ergonomics of a builder API. `seccompiler`'s value is
//!    argument filtering and a JSON policy format, neither of which this
//!    filter uses.
//! 3. **No new supply chain.** `libc` is already a workspace dependency and
//!    already in `Cargo.lock`, so the jail around the project's most
//!    dangerous code path adds no new crate at all.
//!
//! Where this module improves on the `sc-server` precedent: syscall numbers
//! come from `libc::SYS_*` rather than being hardcoded, the architecture is
//! checked before the syscall number is trusted, the x32 ABI is rejected,
//! failures are fatal instead of a warning, and the assembled program's jump
//! offsets are unit-tested.
//!
//! # Fail-closed
//!
//! [`install`] returns `Err` rather than silently doing nothing if it cannot
//! install the policy it promised — including on an architecture this module
//! has no `AUDIT_ARCH` for. The caller ([`super::enter_jail`]) treats that as
//! fatal for the worker child: a worker that could not be jailed must not run
//! a decoder.

#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
pub use filter::install;

#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
use crate::error::PreviewError;

/// Fail-closed stub for architectures this module has no verified
/// `AUDIT_ARCH` / syscall-number mapping for. Refusing to start is the
/// correct behaviour: the alternative is a worker running untrusted
/// decoders with no syscall filter while the docs claim otherwise.
#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
pub fn install() -> Result<(), PreviewError> {
    Err(PreviewError::Worker(
        "no seccomp filter is defined for this architecture; refusing to run an unjailed \
         preview worker"
            .into(),
    ))
}

#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
mod filter {
    use std::io;

    use crate::error::PreviewError;

    // --- classic BPF opcodes (linux/bpf_common.h) ---
    const BPF_LD: u16 = 0x00;
    const BPF_W: u16 = 0x00;
    const BPF_ABS: u16 = 0x20;
    const BPF_JMP: u16 = 0x05;
    const BPF_JEQ: u16 = 0x10;
    #[cfg(target_arch = "x86_64")]
    const BPF_JGE: u16 = 0x30;
    const BPF_K: u16 = 0x00;
    const BPF_RET: u16 = 0x06;

    // --- seccomp (linux/seccomp.h) ---
    const SECCOMP_SET_MODE_FILTER: libc::c_long = 1;
    /// `SECCOMP_MODE_FILTER`, spelled out rather than taken from `libc` so the
    /// pre-3.17 `prctl` fallback below does not depend on which `libc`
    /// version/target happens to export the constant.
    const SECCOMP_MODE_FILTER: libc::c_int = 2;
    const SECCOMP_RET_KILL_PROCESS: u32 = 0x8000_0000;
    const SECCOMP_RET_ALLOW: u32 = 0x7fff_0000;

    // --- `struct seccomp_data` field offsets ---
    const OFF_NR: u32 = 0;
    const OFF_ARCH: u32 = 4;

    /// The arch check is not decoration: without it, a syscall number that
    /// means something harmless under one ABI can mean something else
    /// entirely under another, and the filter would wave it through.
    #[cfg(target_arch = "x86_64")]
    const AUDIT_ARCH: u32 = 0xC000_003E; // AUDIT_ARCH_X86_64
    #[cfg(target_arch = "aarch64")]
    const AUDIT_ARCH: u32 = 0xC000_00B7; // AUDIT_ARCH_AARCH64

    /// On x86-64 the x32 ABI reports the same `arch` value with this bit set
    /// on the syscall number, so an x32 task would otherwise be matched
    /// against x86-64 syscall numbers. Reject the whole range.
    #[cfg(target_arch = "x86_64")]
    const X32_SYSCALL_BIT: u32 = 0x4000_0000;

    #[cfg(target_arch = "x86_64")]
    const PROLOGUE: usize = 4;
    #[cfg(not(target_arch = "x86_64"))]
    const PROLOGUE: usize = 3;

    const fn stmt(code: u16, k: u32) -> libc::sock_filter {
        libc::sock_filter { code, jt: 0, jf: 0, k }
    }

    const fn jump(code: u16, k: u32, jt: u8, jf: u8) -> libc::sock_filter {
        libc::sock_filter { code, jt, jf, k }
    }

    /// The allow-list from `DESIGN-PREVIEW.md` §4.2, plus `recvmsg`/`sendmsg`
    /// (the design doc's pseudo-code omits them, but they are how the worker
    /// receives its job and its fds in the first place — without them the
    /// worker cannot take a single job).
    ///
    /// Deliberately absent, and worth naming because each is a proof
    /// obligation the jail discharges: `openat` (no files by name, ever),
    /// `socket`/`connect`/`sendto` (no network), `clone`/`fork`/`execve` (no
    /// new processes), `ptrace`, `mprotect` (no W^X flips), `statx`.
    ///
    /// `statx` in particular is why the worker reads its input with a manual
    /// `read(2)` loop instead of `Read::read_to_end`: std's `File`
    /// specialization of `read_to_end` sizes the buffer via `File::metadata`,
    /// which on a modern kernel issues `statx`, which this filter kills.
    /// Keeping the list identical to the design doc was worth the four extra
    /// lines in the worker loop.
    fn allowed_syscalls() -> Vec<libc::c_long> {
        vec![
            libc::SYS_read,
            libc::SYS_pread64,
            libc::SYS_write,
            libc::SYS_pwrite64,
            libc::SYS_mmap,
            libc::SYS_munmap,
            libc::SYS_mremap,
            libc::SYS_brk,
            libc::SYS_close,
            libc::SYS_fstat,
            libc::SYS_lseek,
            libc::SYS_futex,
            libc::SYS_sched_yield,
            libc::SYS_exit,
            libc::SYS_exit_group,
            libc::SYS_rt_sigreturn,
            libc::SYS_rt_sigprocmask,
            libc::SYS_madvise,
            libc::SYS_getrandom,
            libc::SYS_clock_gettime,
            libc::SYS_recvmsg,
            libc::SYS_sendmsg,
        ]
    }

    /// Assemble the filter program.
    ///
    /// ```text
    ///   0                    A = seccomp_data.arch
    ///   1                    if A != AUDIT_ARCH        -> KILL
    ///   2                    A = seccomp_data.nr
    ///   3                    if A >= X32_SYSCALL_BIT   -> KILL   (x86_64 only)
    ///   PROLOGUE..+n-1       if A == allowed[i]        -> ALLOW
    ///   PROLOGUE+n           KILL                                (default)
    ///   PROLOGUE+n+1         ALLOW
    /// ```
    ///
    /// BPF jump offsets are unsigned 8-bit and relative to the *following*
    /// instruction, so the longest jump here must stay under 256. With 22
    /// entries there is a lot of headroom, but this checks rather than
    /// trusting that nobody ever grows the list.
    fn build_program(allowed: &[libc::c_long]) -> Result<Vec<libc::sock_filter>, PreviewError> {
        let n = allowed.len();
        let kill_idx = PROLOGUE + n;
        let allow_idx = kill_idx + 1;

        if kill_idx.saturating_sub(2) > u8::MAX as usize {
            return Err(PreviewError::Worker(format!(
                "seccomp allow-list of {n} syscalls does not fit in 8-bit BPF jump offsets"
            )));
        }

        let mut prog = Vec::with_capacity(allow_idx + 1);

        prog.push(stmt(BPF_LD | BPF_W | BPF_ABS, OFF_ARCH));
        prog.push(jump(
            BPF_JMP | BPF_JEQ | BPF_K,
            AUDIT_ARCH,
            0,
            (kill_idx - 2) as u8,
        ));
        prog.push(stmt(BPF_LD | BPF_W | BPF_ABS, OFF_NR));
        #[cfg(target_arch = "x86_64")]
        prog.push(jump(
            BPF_JMP | BPF_JGE | BPF_K,
            X32_SYSCALL_BIT,
            (kill_idx - 4) as u8,
            0,
        ));

        for (i, nr) in allowed.iter().enumerate() {
            let here = PROLOGUE + i;
            prog.push(jump(
                BPF_JMP | BPF_JEQ | BPF_K,
                *nr as u32,
                (allow_idx - here - 1) as u8,
                0,
            ));
        }

        prog.push(stmt(BPF_RET | BPF_K, SECCOMP_RET_KILL_PROCESS));
        prog.push(stmt(BPF_RET | BPF_K, SECCOMP_RET_ALLOW));

        debug_assert_eq!(prog.len(), allow_idx + 1);
        Ok(prog)
    }

    /// Set `PR_SET_NO_NEW_PRIVS`. Required before an unprivileged process may
    /// install a seccomp filter, and required by Landlock too; setting it
    /// here as well as letting the `landlock` crate do it keeps the two
    /// layers independent of each other's ordering.
    fn set_no_new_privs() -> Result<(), PreviewError> {
        // SAFETY: prctl with a fixed, argument-free option.
        let rc = unsafe { libc::prctl(libc::PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) };
        if rc != 0 {
            return Err(PreviewError::Worker(format!(
                "prctl(PR_SET_NO_NEW_PRIVS) failed: {}",
                io::Error::last_os_error()
            )));
        }
        Ok(())
    }

    /// Install the allow-list. Irreversible for the life of the process.
    ///
    /// Must be the **last** thing the jail does: everything after this call
    /// is restricted to the twenty-two syscalls above.
    pub fn install() -> Result<(), PreviewError> {
        set_no_new_privs()?;

        let mut prog = build_program(&allowed_syscalls())?;
        let fprog = libc::sock_fprog {
            len: prog.len() as libc::c_ushort,
            filter: prog.as_mut_ptr(),
        };

        // SAFETY: `fprog` points at a live, correctly-sized program for the
        // duration of the call; the kernel reads it and does not retain the
        // pointer.
        let rc = unsafe {
            libc::syscall(
                libc::SYS_seccomp,
                SECCOMP_SET_MODE_FILTER,
                0 as libc::c_long,
                &fprog as *const libc::sock_fprog,
            )
        };
        if rc == 0 {
            return Ok(());
        }

        let err = io::Error::last_os_error();
        if err.raw_os_error() == Some(libc::ENOSYS) {
            // Pre-3.17 kernels only have the prctl entry point. Same filter,
            // same semantics for a single-threaded process.
            // SAFETY: as above.
            let rc = unsafe {
                libc::prctl(
                    libc::PR_SET_SECCOMP,
                    SECCOMP_MODE_FILTER,
                    &fprog as *const libc::sock_fprog,
                    0,
                    0,
                )
            };
            if rc == 0 {
                return Ok(());
            }
            return Err(PreviewError::Worker(format!(
                "prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER) failed: {}",
                io::Error::last_os_error()
            )));
        }

        Err(PreviewError::Worker(format!(
            "seccomp(SECCOMP_SET_MODE_FILTER) failed: {err}"
        )))
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        /// The program has to be well-formed *before* it reaches the kernel:
        /// a bad jump offset is a filter that silently allows something.
        #[test]
        fn program_is_well_formed_and_every_jump_lands_in_range() {
            let prog = build_program(&allowed_syscalls()).unwrap();
            let n = prog.len();
            assert!(n > allowed_syscalls().len());

            assert_eq!(prog[n - 2].code, BPF_RET | BPF_K);
            assert_eq!(prog[n - 2].k, SECCOMP_RET_KILL_PROCESS);
            assert_eq!(prog[n - 1].code, BPF_RET | BPF_K);
            assert_eq!(prog[n - 1].k, SECCOMP_RET_ALLOW);

            for (i, ins) in prog.iter().enumerate() {
                if ins.code & BPF_JMP != BPF_JMP {
                    continue;
                }
                for off in [ins.jt, ins.jf] {
                    let target = i + 1 + off as usize;
                    assert!(target < n, "instruction {i} jumps out of the program");
                }
            }
        }

        #[test]
        fn arch_is_checked_before_the_syscall_number_is_ever_loaded() {
            let prog = build_program(&allowed_syscalls()).unwrap();
            assert_eq!(prog[0].code, BPF_LD | BPF_W | BPF_ABS);
            assert_eq!(prog[0].k, OFF_ARCH);
            assert_eq!(prog[1].k, AUDIT_ARCH);
            // A mismatched arch must land exactly on KILL, not on ALLOW.
            let target = 1 + 1 + prog[1].jf as usize;
            assert_eq!(prog[target].k, SECCOMP_RET_KILL_PROCESS);
        }

        #[cfg(target_arch = "x86_64")]
        #[test]
        fn x32_syscalls_are_killed_rather_than_matched_against_x86_64_numbers() {
            let prog = build_program(&allowed_syscalls()).unwrap();
            assert_eq!(prog[3].code, BPF_JMP | BPF_JGE | BPF_K);
            assert_eq!(prog[3].k, X32_SYSCALL_BIT);
            let target = 3 + 1 + prog[3].jt as usize;
            assert_eq!(prog[target].k, SECCOMP_RET_KILL_PROCESS);
        }

        #[test]
        fn each_allowed_syscall_jumps_to_allow_and_a_miss_falls_through() {
            let allowed = allowed_syscalls();
            let prog = build_program(&allowed).unwrap();
            let allow_idx = prog.len() - 1;

            for (i, nr) in allowed.iter().enumerate() {
                let ins = prog[PROLOGUE + i];
                assert_eq!(ins.k, *nr as u32);
                assert_eq!(PROLOGUE + i + 1 + ins.jt as usize, allow_idx);
                assert_eq!(ins.jf, 0, "a non-match must fall through, never jump");
            }
        }

        /// The syscalls the jail exists to deny. If one of these ever appears
        /// in the allow-list, the proofs in `examples/jail_proof.rs` stop
        /// meaning what they say they mean.
        #[test]
        fn the_denied_syscalls_are_actually_absent() {
            let allowed = allowed_syscalls();
            for (name, nr) in [
                ("openat", libc::SYS_openat),
                ("socket", libc::SYS_socket),
                ("connect", libc::SYS_connect),
                ("clone", libc::SYS_clone),
                ("execve", libc::SYS_execve),
                ("ptrace", libc::SYS_ptrace),
                ("mprotect", libc::SYS_mprotect),
                ("statx", libc::SYS_statx),
            ] {
                assert!(!allowed.contains(&nr), "{name} must not be allowed");
            }
        }
    }
}

File: go/internal/jail/seccomp.go
//go:build linux

package jail

import (
	"fmt"
	"runtime"
	"unsafe"

	"github.com/heavycaffeiner/stowcloud/go/internal/num"
	"golang.org/x/sys/unix"
)

// Classic BPF opcodes, from linux/bpf_common.h. Each is spelled as the OR of
// its named parts so it can be checked against the header by eye; several of
// those parts are genuinely zero there.
const (
	bpfLd  = 0x00
	bpfW   = 0x00
	bpfAbs = 0x20
	bpfJmp = 0x05
	bpfJeq = 0x10
	bpfJge = 0x30
	bpfK   = 0x00
	bpfRet = 0x06

	opLoad      = bpfLd | bpfW | bpfAbs
	opJumpEqual = bpfJmp | bpfJeq | bpfK
	opJumpGE    = bpfJmp | bpfJge | bpfK
	opReturn    = bpfRet | bpfK
)

// Field offsets in struct seccomp_data. arch is at 4 and the syscall number is
// at 0, and reading them in that order is the whole of D4.
const (
	offNr   = 0
	offArch = 4
)

// The two AUDIT_ARCH values this build has a verified syscall mapping for.
const (
	auditArchAmd64 = 0xC000003E
	auditArchArm64 = 0xC00000B7

	// On x86-64 the x32 ABI reports the same AUDIT_ARCH with this bit set on
	// the syscall number, so an x32 task would otherwise be matched against
	// x86-64 numbers. The whole range is rejected.
	x32SyscallBit = 0x40000000
)

// FilterKind selects one of the two policies. Both are flat syscall number
// lists with no argument inspection, which is what keeps each readable in one
// screen.
type FilterKind uint8

const (
	// FilterProcess is the server's deny-list. It returns EPERM rather than
	// killing, because it is a second line of defence beside Landlock and an
	// unprivileged uid, not a sandbox on its own.
	FilterProcess FilterKind = iota

	// FilterWorker is the decoder's allow-list. Everything not on it kills the
	// process, which is the designed outcome: a crafted input that kills a
	// decoder costs one thumbnail.
	FilterWorker

	// FilterWorkerAudit is the allow-list with the kill replaced by a log
	// record. It is how the shipped list is measured rather than guessed:
	// the worker runs under it against a corpus of real images, and the audit
	// log names every syscall the list is missing.
	//
	// It is never shipped. A filter that logs instead of killing is not a
	// sandbox, and the one caller is the measurement command.
	FilterWorkerAudit

	// FilterWorkerTrap is the allow-list with the kill replaced by SIGSYS.
	//
	// It exists because SECCOMP_RET_LOG writes to the kernel audit log, which
	// needs root to read: on a machine where auditd owns it, a measurement run
	// as an ordinary user gets a clean result that means only that it could not
	// see the records. A trap is delivered to the process itself, so the
	// measurement records what it observed rather than what it was permitted to
	// read, and it needs no privilege.
	//
	// Never shipped either, for the same reason: a trap a handler swallows is
	// not a sandbox.
	FilterWorkerTrap
)

// archProfileFor is a table rather than a build tag, because an architecture
// with no entry has to be a refusal at runtime and not a compile error nobody
// sees until they try to build for it. Taking the name as an argument is what
// lets the refusal be tested from the two architectures that do have entries,
// which is otherwise a branch nothing ever runs.
func archProfileFor(goarch string) (auditArch uint32, rejectX32 bool, ok bool) {
	switch goarch {
	case "amd64":
		return auditArchAmd64, true, true
	case "arm64":
		return auditArchArm64, false, true
	}
	return 0, false, false
}

func archProfile() (auditArch uint32, rejectX32 bool, ok bool) {
	return archProfileFor(runtime.GOARCH)
}

// deniedSyscalls is the server's list. Each one is a way out of the process or
// into another one, and none of them has a legitimate caller here.
func deniedSyscalls() []int {
	return []int{
		unix.SYS_PTRACE,
		unix.SYS_PROCESS_VM_READV,
		unix.SYS_PROCESS_VM_WRITEV,
		unix.SYS_MOUNT,
		unix.SYS_KEXEC_LOAD,
		unix.SYS_KEXEC_FILE_LOAD,
		unix.SYS_BPF,
		unix.SYS_USERFAULTFD,
	}
}

// allowedSyscalls is the decoder's list.
//
// What is absent is the proof obligation, and each absence is deliberate: no
// openat and no openat2, so no file by name ever; no socket and no connect, so
// no network; no clone and no execve, so no new process; no ptrace. recvmsg and
// sendmsg are on it because they are how a job and its two descriptors arrive
// at all.
//
// This list is measured, not reasoned about. A real worker was driven over a
// real socket against the committed corpus and traced; the union of its steady
// state and the decode phase is thirteen calls, and every one of them is here.
// The remainder are startup and teardown the trace showed before the filter is
// installed or on the way out, kept because a decode that grows the heap or
// takes a signal must not be killed for it.
//
// Three came from the measurement rather than the estimate, and none would
// have been guessed:
//
//   - fcntl, which os.NewFile issues as F_GETFL on each descriptor a job
//     arrives with, to learn whether it is readable or writable.
//   - epoll_pwait and getpid, which the Go runtime's scheduler and its
//     signal path use even with one thread.
//   - epoll_create1, eventfd2 and epoll_ctl, which are the network poller
//     being built. The runtime does that lazily, on whichever job first parks
//     a goroutine on the socket, so a worker survives its first few and then
//     is killed. The original measurement had epoll_pwait without them, which
//     is the wait without the setup: it had captured a poller that was already
//     running before the trace began.
//
// clone is deliberately absent and the measurement is what makes that
// affordable. Every clone the trace showed carried CLONE_THREAD, so it was the
// runtime adding an OS thread rather than a fork. A worker that cannot clone
// cannot fork either, which is the property this list is for.
//
// GOMAXPROCS=1 does not mean one OS thread, and an earlier version of this
// note claimed it did. It bounds goroutines running Go code, not threads: the
// runtime still has five or so for its own work. What makes clone's absence
// hold is that none of them is started after the filter is installed, not that
// they do not exist.
func allowedSyscalls() []int {
	return []int{
		unix.SYS_READ,
		unix.SYS_PREAD64,
		unix.SYS_WRITE,
		unix.SYS_PWRITE64,
		unix.SYS_MMAP,
		unix.SYS_MUNMAP,
		unix.SYS_MREMAP,
		unix.SYS_BRK,
		unix.SYS_CLOSE,
		unix.SYS_FSTAT,
		unix.SYS_LSEEK,
		unix.SYS_FUTEX,
		unix.SYS_SCHED_YIELD,
		unix.SYS_EXIT,
		unix.SYS_EXIT_GROUP,
		unix.SYS_RT_SIGRETURN,
		unix.SYS_RT_SIGPROCMASK,
		unix.SYS_MADVISE,
		unix.SYS_GETRANDOM,
		unix.SYS_CLOCK_GETTIME,
		unix.SYS_RECVMSG,
		unix.SYS_SENDMSG,
		unix.SYS_NANOSLEEP,
		unix.SYS_SIGALTSTACK,
		unix.SYS_RT_SIGACTION,
		unix.SYS_GETTID,
		unix.SYS_TGKILL,

		// Measured, and absent from the estimate. See the note above.
		unix.SYS_FCNTL,
		unix.SYS_EPOLL_PWAIT,
		unix.SYS_GETPID,
		unix.SYS_EPOLL_CREATE1,
		unix.SYS_EVENTFD2,
		unix.SYS_EPOLL_CTL,

		// prctl, which the runtime issues as PR_SET_VMA_ANON_NAME to label its
		// own mappings so they are identifiable in /proc/pid/maps. It is on by
		// default from Go 1.26 and happens whenever the heap grows, so a
		// worker decoding a larger image than the last one is killed for it.
		//
		// It grants nothing: PR_SET_VMA only renames a mapping this process
		// already owns.
		unix.SYS_PRCTL,
	}
}

func stmt(code uint16, k uint32) unix.SockFilter {
	return unix.SockFilter{Code: code, K: k}
}

func jump(code uint16, k uint32, jt, jf uint8) unix.SockFilter {
	return unix.SockFilter{Code: code, Jt: jt, Jf: jf, K: k}
}

// assemble builds the program.
//
//	0                  A = seccomp_data.arch
//	1                  if A != AUDIT_ARCH      -> kill
//	2                  A = seccomp_data.nr
//	3                  if A >= X32_SYSCALL_BIT -> kill   (amd64 only)
//	prologue..+n-1     one compare per listed syscall
//	prologue+n         the default action for this policy
//	prologue+n+1       the matched action for this policy
//	prologue+n+2       kill, where an unexpected ABI lands
//
// The arch check comes first and it is not decoration. A syscall number is only
// meaningful together with the ABI that issued it, so a filter matching numbers
// without pinning the arch is matching numbers that mean something else.
//
// The refusal is its own instruction rather than the policy's default action,
// and that is the difference between the two things this fixes. The server's
// default action is ALLOW, so pointing an ABI mismatch at it would wave through
// exactly the task the check exists to catch. An unexpected ABI is killed under
// both policies.
//
// BPF jump offsets are unsigned 8-bit and relative to the following
// instruction, so the longest jump has to stay under 256. There is headroom at
// these list lengths; this checks rather than trusting that nobody grows one.
func assemble(kind FilterKind) ([]unix.SockFilter, error) {
	return assembleFor(kind, runtime.GOARCH)
}

func assembleFor(kind FilterKind, goarch string) ([]unix.SockFilter, error) {
	auditArch, rejectX32, ok := archProfileFor(goarch)
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrArchUnsupported, goarch)
	}

	var list []int
	var matched, unmatched uint32
	switch kind {
	case FilterProcess:
		// A listed number is denied and everything else runs.
		list = deniedSyscalls()
		matched = unix.SECCOMP_RET_ERRNO | uint32(unix.EPERM)
		unmatched = unix.SECCOMP_RET_ALLOW
	case FilterWorker:
		// A listed number runs and everything else kills the process.
		list = allowedSyscalls()
		matched = unix.SECCOMP_RET_ALLOW
		unmatched = unix.SECCOMP_RET_KILL_PROCESS
	case FilterWorkerAudit:
		// The same list, but an unlisted call is logged and then allowed, so
		// one run finds every missing entry rather than the first.
		list = allowedSyscalls()
		matched = unix.SECCOMP_RET_ALLOW
		unmatched = unix.SECCOMP_RET_LOG
	case FilterWorkerTrap:
		// The same list, with an unlisted call raising SIGSYS in the calling
		// thread. The handler records the number and the run continues.
		list = allowedSyscalls()
		matched = unix.SECCOMP_RET_ALLOW
		unmatched = unix.SECCOMP_RET_TRAP
	default:
		return nil, fmt.Errorf("unknown seccomp filter kind %d", kind)
	}

	prologue := 3
	if rejectX32 {
		prologue = 4
	}
	n := len(list)
	unmatchedIdx := prologue + n
	matchedIdx := unmatchedIdx + 1
	refuseIdx := matchedIdx + 1

	// Every jump targets one of the three terminal instructions, so the longest
	// one is the reach from the first jump to the last instruction.
	if refuseIdx > 1+int(^uint8(0)) {
		return nil, fmt.Errorf("a %d entry seccomp list does not fit in 8-bit BPF jump offsets", n)
	}

	prog := make([]unix.SockFilter, 0, refuseIdx+1)
	prog = append(prog, stmt(opLoad, offArch))

	archMiss, err := offsetTo(refuseIdx, 1)
	if err != nil {
		return nil, err
	}
	prog = append(prog, jump(opJumpEqual, auditArch, 0, archMiss))
	prog = append(prog, stmt(opLoad, offNr))

	if rejectX32 {
		x32, err := offsetTo(refuseIdx, 3)
		if err != nil {
			return nil, err
		}
		prog = append(prog, jump(opJumpGE, x32SyscallBit, x32, 0))
	}

	for i, nr := range list {
		here := prologue + i
		off, err := offsetTo(matchedIdx, here)
		if err != nil {
			return nil, err
		}
		number, err := num.Narrow[uint32](nr)
		if err != nil {
			return nil, fmt.Errorf("syscall number %d: %w", nr, err)
		}
		// A non-match falls through to the next compare, never jumps.
		prog = append(prog, jump(opJumpEqual, number, off, 0))
	}

	prog = append(prog, stmt(opReturn, unmatched))
	prog = append(prog, stmt(opReturn, matched))
	prog = append(prog, stmt(opReturn, unix.SECCOMP_RET_KILL_PROCESS))
	if len(prog) != refuseIdx+1 {
		return nil, fmt.Errorf("assembled %d instructions, expected %d", len(prog), refuseIdx+1)
	}
	return prog, nil
}

// offsetTo is the jump distance from the instruction at from to the one at to,
// which BPF counts from the instruction after the jump.
func offsetTo(to, from int) (uint8, error) {
	d := to - from - 1
	if d < 0 {
		return 0, fmt.Errorf("a backward BPF jump from %d to %d, which the verifier refuses", from, to)
	}
	return num.Narrow[uint8](d)
}

// InstallSeccomp assembles and installs the policy for kind with TSYNC, so it
// covers every thread the Go runtime has already started rather than whichever
// one this goroutine happens to be on.
//
// Irreversible for the life of the process, and for the worker it has to be the
// last thing the jail does.
func InstallSeccomp(kind FilterKind) error {
	prog, err := assemble(kind)
	if err != nil {
		return err
	}

	// Required before an unprivileged process may install a filter at all.
	if perr := unix.Prctl(unix.PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0); perr != nil {
		return fmt.Errorf("prctl(PR_SET_NO_NEW_PRIVS): %w", perr)
	}

	length, err := num.Narrow[uint16](len(prog))
	if err != nil {
		return fmt.Errorf("seccomp program length: %w", err)
	}
	fprog := unix.SockFprog{Len: length, Filter: &prog[0]}

	// TSYNC applies the filter to every thread or fails atomically, which is
	// the answer Landlock has no equivalent of.
	_, _, errno := unix.Syscall(unix.SYS_SECCOMP,
		uintptr(unix.SECCOMP_SET_MODE_FILTER),
		uintptr(unix.SECCOMP_FILTER_FLAG_TSYNC),
		//nolint:gosec // the kernel takes a sock_fprog pointer; there is no wrapper for seccomp.
		uintptr(unsafe.Pointer(&fprog)))
	runtime.KeepAlive(prog)
	if errno != 0 {
		return fmt.Errorf("seccomp(SECCOMP_SET_MODE_FILTER, TSYNC): %w", errno)
	}
	return nil
}

// AllowedSyscalls is the worker's list, for the command that measures it.
//
// Exported so the audit run and the list it checks cannot drift apart: a
// measurement against a copy of the list proves nothing about the one that
// ships.
func AllowedSyscalls() []int { return allowedSyscalls() }
